### PR TITLE
Makefile.defines: Enable more build warnings and add option to enable  -Werror

### PR DESCRIPTION
**Pull request by xchapron-ledger**
## Description

Enable more build warnings and add option to enable ` -Werror`.
List inspired from app usage, discussion with other devs and https://embeddedartistry.com/blog/2017/06/07/warnings-weverything-and-the-kitchen-sink/ and https://clang.llvm.org/docs/DiagnosticsReference.html#id413

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Tests
- [ ] Documentation
- [x] Other (for changes that might not fit in any category)

### CODE REVIEW

**Comment by xchapron-ledger** [comment]
Adding `-Wimplicit-fallthrough` enforce using either clang recognise marker: `__attribute__ ((fallthrough))` or SDK define: `FALL_THROUGH`.
This means this should be changed:
https://github.com/LedgerHQ/app-boilerplate/blob/master/src/io.c#L56

File: include/decorators.h [hunk 1]
```diff
@@ -55,6 +55,6 @@
   #if defined(__GNUC__) && __GNUC__ >= 7 || defined(__clang__) && __clang_major__ >= 12
```
Comment (fbeutin-ledger) [inline include/decorators.h:55]: To remove ?
Maybe assert and refuse to compile if too old. I'm not sure the compilation would proceed successfully we've never tried and we don't want to support it

File: Makefile.defines [hunk 2]
```diff
@@ -110,7 +110,21 @@ CFLAGS   += $(OPTI_LVL) $(DBG_LVL)
...
+
+CFLAGS   += -Wall -Wextra
+CFLAGS   += -Wno-main
+CFLAGS   += -Werror=int-to-pointer-cast
+
+# Additional Clang warnings
+CFLAGS   += -Wno-error=int-conversion -Wimplicit-fallthrough
+CFLAGS   += -Wvla -Wundef -Wshadow -Wformat=2 -Wformat-security -Wwrite-strings
+
+WERROR = 0
```
Comment (fbeutin-ledger) [inline Makefile.defines:123]: See my comment on embedded-apps-core :)
I propose `?=`

File: include/decorators.h (hunk 1 above)
Comment (xchapron-ledger) [reply include/decorators.h:55]: As discussed, this will be addressed in another task / PR

**Review by fbeutin-ledger — APPROVED** [review]

**Review by tjoly-ledger — APPROVED** [review]
bolos_ng: compilation OK
OSU: compilation OK
OSU: test OK

**Comment by xchapron-ledger** [comment]
This break the build of app-fio due to the fact WERROR is not null: https://github.com/LedgerHQ/app-fio/blob/develop/Makefile#L127
And the app have warning when building:

```
[CC]   build/nanos/obj/eos_utils.o
src/eos_utils.c:178:17: warning: variable length array used [-Wvla]
    uint8_t buf[size];
                ^~~~
1 warning generated.
[CC]   build/nanos/obj/fio.o
[CC]   build/nanos/obj/getPublicKey.o
[CC]   build/nanos/obj/getSerial.o
src/getSerial.c:22:22: warning: variable length array used [-Wvla]
    uint8_t response[SERIAL_LENGTH];
                     ^~~~~~~~~~~~~
1 warning generated.
[CC]   build/nanos/obj/getVersion.o
[CC]   build/nanos/obj/glyphs.o
[CC]   build/nanos/obj/handlers.o
[CC]   build/nanos/obj/hexUtils.o
src/hexUtils.c:89:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"ff", 0xff},
         ^~~~
src/hexUtils.c:90:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"00", 0x00},
         ^~~~
src/hexUtils.c:91:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"1a", 0x1a},
         ^~~~
src/hexUtils.c:92:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"2b", 0x2b},
         ^~~~
src/hexUtils.c:93:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"3c", 0x3c},
         ^~~~
src/hexUtils.c:94:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"4d", 0x4d},
         ^~~~
src/hexUtils.c:95:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"5f", 0x5f},
         ^~~~
src/hexUtils.c:96:10: error: initializing 'char *' with an expression of type 'const char [3]' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        {"98", 0x98},
         ^~~~
8 errors generated.
```

**Pull request merged**